File: Lessons/7 StacksandQues/Fish/fish.cpp
#include <stack>
int solution(vector<int> &A, vector<int> &B) {
  stack<int> upstream;
  int eaten = 0;
  int fish=0;
  for (auto i = 0; i < A.size(); i++) {
    if (B[i] == 1) {
      upstream.push(A[i]);
    } else {
      while (upstream.size() > 0) {
        eaten += 1;
        fish = upstream.top();
         upstream.pop();
        if (fish > A[i]) {
          upstream.push(fish);
          break;
        }
      }
    }
  }
  return A.size() - eaten;
}

File: Lessons/7 StacksandQues/Fish/fish.java
import java.util.Arrays;
import java.util.Stack;

class Solution {
  public int solution(int[] A, int[] B) {
    if (A.length == 0) {
      return 0;
    }
    Stack<Integer> upstream = new Stack<Integer>();
    int eaten = 0;
    int fish = 0;

    for (int i = 0; i < A.length; i++) {
      if (B[i] == 1) {
        upstream.push(A[i]);
      } else {
        while (!upstream.empty()) {
          eaten += 1;
          fish = upstream.pop();
          if (fish > A[i]) {
            upstream.push(fish);
            break;
          }
        }
      }
    }
    return (A.length - eaten);
  }
}
